src/internal/ans.ts

```diff
@@ -147,29 +149,359 @@ export async function registerName(args: RegisterNameParameters): Promise<InputS
+    aptosConfig,
+    query: {
+      query: GetNames,
+      variables: {
+        where_condition: where,
+        limit: 1,
+      },
+    },
+    originMethod: "getName",
+  });
+
+  return data.current_aptos_names[0] as GetANSNameResponse[0] | undefined;
+}
+
+export type GetNamesArgs = {
+  options?: {
+    pagination?: PaginationArgs;
+    orderBy?: OrderBy<GetANSNameResponse[0]>;
+    where?: CurrentAptosNamesBoolExp;
+  };
+} & (
+  | { query: "owner"; ownerAddress: AccountAddressInput }
+  | { query: "owner:domains"; ownerAddress: AccountAddressInput }
+  | { query: "owner:subdomains"; ownerAddress: AccountAddressInput }
+  | { query: "domain:subdomains"; domain: string }
+);
+
+export async function getNames(args: { aptosConfig: AptosConfig } & GetNamesArgs): Promise<GetANSNameResponse> {
+  if (!args.query) throw new Error("Must provide a query type");
+
+  const { aptosConfig, query, options } = args;
+
+  /**
+   * Names that are past expiration but within the grace period should be
+   * included in this query. Names that expire outside the range of the grace
+   * period should be excluded. We calculate this by fetching the grace
+   * period from the chain, subtract it from the current date, and fetch all
+   * names that expire after this new date.
+   */
+  const gracePeriodInSeconds = await getGracePeriodInSeconds({ aptosConfig });
+  const gracePeriodInDays = gracePeriodInSeconds / 60 / 60 / 24;
+  const now = () => new Date();
+  const cutoff = new Date(now().setDate(now().getDate() - gracePeriodInDays)).toISOString();
+
+  let where: CurrentAptosNamesBoolExp;
+  switch (query) {
+    case "owner":
+      where = {
+        owner_address: { _eq: args.ownerAddress.toString() },
+        expiration_timestamp: { _gte: cutoff },
+      };
+      break;
+
+    case "owner:domains":
+      where = {
+        owner_address: { _eq: args.ownerAddress.toString() },
+        expiration_timestamp: { _gte: cutoff },
+        subdomain: { _eq: "" },
+      };
+      break;
+
+    case "owner:subdomains":
+      where = {
+        owner_address: { _eq: args.ownerAddress.toString() },
+        expiration_timestamp: { _gte: cutoff },
+        subdomain: { _neq: "" },
+      };
+      break;
+
+    case "domain:subdomains":
+      where = {
+        domain: { _eq: args.domain },
+        subdomain: { _neq: "" },
+        // Note: Subdomains have a different expiration policy than domains, no
+        // need to filter on expiration_timestamp
+      };
+      break;
+
+    default:
+      throw new Error(`Invalid query type: ${query}`);
+  }
+
+  const data = await queryIndexer<GetNamesQuery>({
+    aptosConfig,
+    originMethod: "getNames",
+    query: {
+      query: GetNames,
+      variables: {
+        limit: options?.pagination?.limit,
+        offset: options?.pagination?.offset,
+        order_by: options?.orderBy,
+        where_condition: { ...where, ...(args.options?.where ?? {}) },
```

Comment:
I think you would want to switch order so your "fixed where" will not get overridden by the user input

```suggestion
        where_condition: { ...(args.options?.where ?? {}), ...where },
```

Reply:
Done!